src/common/static_ptr.h

```diff
@@ -0,0 +1,395 @@
+    r.operate = p.operate;
+  }
+  return r;
+}
+template<typename U, std::size_t Z, typename T, std::size_t S>
+static_ptr<U, Z> static_pointer_cast(static_ptr<T, S>&& p) {
+  static_assert(Z >= S,
+                "Value too large.");
+  static_ptr<U, Z> r;
+  if (static_cast<U*>(p.get())) {
+    p.operate(_mem::op::move, &p.buf, &r.buf);
+    r.operate = p.operate;
+  }
+  return r;
+}
+
+// Here the conditional is actually important and ensures we have the
+// same behavior as dynamic_cast.
+//
+template<typename U, std::size_t Z, typename T, std::size_t S>
+static_ptr<U, Z> dynamic_pointer_cast(const static_ptr<T, S>& p) {
+  static_assert(Z >= S,
+                "Value too large.");
+  static_ptr<U, Z> r;
+  if (dynamic_cast<U*>(p.get())) {
+    p.operate(_mem::op::copy, &p.buf, &r.buf);
+    r.operate = p.operate;
+  }
+  return r;
+}
+template<typename U, std::size_t Z, typename T, std::size_t S>
+static_ptr<U, Z> dynamic_pointer_cast(static_ptr<T, S>&& p) {
+  static_assert(Z >= S,
+                "Value too large.");
+  static_ptr<U, Z> r;
+  if (dynamic_cast<U*>(p.get())) {
+    p.operate(_mem::op::move, &p.buf, &r.buf);
+    r.operate = p.operate;
+  }
+  return r;
+}
+
+template<typename U, std::size_t Z, typename T, std::size_t S>
+static_ptr<U, Z> const_pointer_cast(const static_ptr<T, S>& p) {
+  static_assert(Z >= S,
+                "Value too large.");
+  static_ptr<U, Z> r;
+  if (const_cast<U*>(p.get())) {
+    p.operate(_mem::op::copy, &p.buf, &r.buf);
+    r.operate = p.operate;
+  }
+  return r;
+}
+template<typename U, std::size_t Z, typename T, std::size_t S>
+static_ptr<U, Z> const_pointer_cast(static_ptr<T, S>&& p) {
+  static_assert(Z >= S,
+                "Value too large.");
+  static_ptr<U, Z> r;
+  if (const_cast<U*>(p.get())) {
+    p.operate(_mem::op::move, &p.buf, &r.buf);
+    r.operate = p.operate;
+  }
+  return r;
+}
+
+// I'm not sure if anyone will ever use this. I can imagine situations
+// where they might. It works, though!
+//
+template<typename U, std::size_t Z, typename T, std::size_t S>
+static_ptr<U, Z> reinterpret_pointer_cast(const static_ptr<T, S>& p) {
+  static_assert(Z >= S,
+                "Value too large.");
+  static_ptr<U, Z> r;
+  p.operate(_mem::op::copy, &p.buf, &r.buf);
+  r.operate = p.operate;
+  return r;
+}
+template<typename U, std::size_t Z, typename T, std::size_t S>
+static_ptr<U, Z> reinterpret_pointer_cast(static_ptr<T, S>&& p) {
+  static_assert(Z >= S,
+                "Value too large.");
+  static_ptr<U, Z> r;
+  p.operate(_mem::op::move, &p.buf, &r.buf);
+  r.operate = p.operate;
+  return r;
+}
+
+// This is the only way to move from a bigger static pointer into a
+// smaller static pointer. The size of the total data stored in the
+// pointer is checked at runtime and if the destination size is large
+// enough, we copy it over.
+//
+// I follow cast semantics. Since this is a pointer-like type, it
+// returns a null value rather than throwing.
```

Comment:
what is this used for?

Reply:
If you have a static_ptr<Base, 4096> in some part of the code and you have a static_ptr<Base, 256> in some other part of the code.

(Maybe the large derivative are only used on some call paths and you don't want to bloat every structure everywhere)

If you know that the actual thing inside the 4096-byte static pointer is only 64 bytes long, resize_pointer_cast will check that it actually IS small enough to fit and copy it. (If it's not small enough to fit, it will just return a null value, similar to dynamic_cast.)

In normal use we just check the declared size at compile time and refuse to copy/move from a bigger static_ptr into a smaller one. This is an escape hatch where you can request to have it checked and done.